### PR TITLE
Refactor gradient APIs for using partials 

### PR DESCRIPTION
**Pull request by rauletorresc**
**Context:** The function can be used with and without specifying the function.

**Description of the Change:** Use `functools.partial` for when the function is `None`

**Benefits:** Favors decorator syntax.

[sc-60279]

### CODE REVIEW

File: frontend/test/pytest/test_gradient.py [hunk 1]
```diff
@@ -1230,5 +1230,24 @@ def cir(x: float):
...
+        """Test usage patterns of catalyst.grad."""
+
+        def fn(x):
+            return x**2
+
+        x = 4.0
+
+        res_pattern_fn_as_argument = grad(fn)(x)
+        res_pattern_partial = grad()(fn)(x)
+        expected = jax.grad(fn)(x)
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_gradient.py:1246]: Mmm... I know this is in the ticket, but if feels weird that `grad` does not take any values whatsoever. I don't like this and I will defer to @dime10 to decide on whether this is a good UI.

**Review by dime10 — APPROVED** [review]
Nice!

File: frontend/test/pytest/test_gradient.py (hunk 1 above)
Comment (dime10) [reply frontend/test/pytest/test_gradient.py:1246]: It's to enable this pattern:
```py
@grad(method="fd")
def f(x):
  ...
  
f(x)
```

I think Raul is just omitting the decorator syntax (that is `grad(method="fd")(fn)(x)` is equivalent to what I wrote), and also left the keyword parameters at their default value, but otherwise they are testing the same thing. 

Admittedly using some keyword arguments would make the test clearer and would also test that capturing the kwargs via `locals` is working properly.
